### DISCUSSION TITLE
Using zaptest and observer

### DISCUSSION THREAD

**Discussion by 3vilhamster** [uber-go/zap#1420] [Q&A]
Hey folks.
I use zap and for test usually I use zaptest to ensure that I'll see logs only from failed tests in output.
But now I need to verify that some logs were written, in this case zap provides a zaptest/observer package. The problem for me is that observer could be used only with zap.New(core) which still writes stdout/stderr together with observer.
I'd like to know if there is a solution that allows me use both zaptest and zaptest/observer together?
Or maybe request a feature/help with PR to do that.

**Comment by official-burak** [comment]
`zap.New(observedCore)` does not write stdout. Observer only records. Console output means another core is in the mix.

To keep zaptest's "print only on fail" writer and capture logs:

```go
obs, logs := observer.New(zap.DebugLevel)
logger := zaptest.NewLogger(t, zaptest.WrapOptions(
	zap.WrapCore(func(c zapcore.Core) zapcore.Core {
		return zapcore.NewTee(c, obs)
	}),
))
```

If you only need assertions, skip zaptest:

```go
obs, logs := observer.New(zap.InfoLevel)
logger := zap.New(obs)
```
